tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,438 @@
+            var samplingPriority = context.TraceContext?.SamplingPriority ?? context.SamplingPriority ?? SamplingPriorityValues.AutoKeep;
             var sampled = samplingPriority > 0 ? "01" : "00";
-            carrierSetter.Set(carrier, TraceParent, $"00-{traceId}-{spanId}-{sampled}");
+
+            return $"00-{traceId}-{spanId}-{sampled}";
         }
 
-        public bool TryExtract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter, out SpanContext? spanContext)
-            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        internal static string CreateTraceStateHeader(SpanContext context)
         {
-            spanContext = null;
+            var samplingPriority = SamplingPriorityToString(context.TraceContext?.SamplingPriority ?? context.SamplingPriority);
+            StringBuilder? sb = null;
 
-            var traceParent = ParseUtility.ParseString(carrier, carrierGetter, TraceParent)?.Trim();
-            if (!string.IsNullOrEmpty(traceParent))
+            try
             {
-                // We found a trace parent (we are reading from the Http Headers)
+                sb = StringBuilderCache.Acquire(100);
+                sb.Append("dd=");
 
-                /* (https://www.w3.org/TR/trace-context/)
-                    Valid traceparent when caller sampled this request:
+                if (samplingPriority != null)
+                {
+                    sb.Append("s:").Append(samplingPriority).Append(';');
+                }
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 01  // sampled
-                    Valid traceparent when caller didn’t sample this request:
+                if (!string.IsNullOrWhiteSpace(context.Origin))
+                {
+                    sb.Append("o:").Append(context.Origin).Append(';');
+                }
+
+                if (context.TraceContext?.Tags?.ToArray() is { Length: > 0 } tags)
+                {
+                    foreach (var tag in tags)
+                    {
+                        if (tag.Key.StartsWith("_dd.p.", StringComparison.Ordinal))
+                        {
+#if NETCOREAPP
+                            var key = tag.Key.AsSpan(start: 6);
+#else
+                            var key = tag.Key.Substring(startIndex: 6);
+#endif
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-00
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 00  // not sampled
-                 */
+                            sb.Append("t.").Append(key).Append(':').Append(tag.Value).Append(';');
```

Comment:
You may need to do some cleanup on `tag.Value` for invalid characters. For readability, you could do an inline call here like `.Append(SanitizePropagatedTagValue(tag.Value))`

Reply:
I added cleanup with both the key and value as per our W3C headers RFC, although I still have concerns about the behavior being different than the `x-datadog-tags` RFC.